python/ray/serve/tests/test_http_state.py

```diff
@@ -3,172 +3,517 @@
+def test_http_proxy_state_update_healthy_check_health_succeed():
+    """Test calling update method on HTTPProxyState when the proxy state is HEALTHY and
+    when the check_health call succeeded
+
+    The proxy state started with HEALTHY. After update is called and ready call
+    succeeded, the status will change to HEALTHY. After the next period of check_health
+    call, the status should stay as HEALTHY.
+    """
+    proxy_state = _create_http_proxy_state()
+
+    # Continuously trigger update. The status should change from STARTING to HEALTHY
+    # when ready.
+    wait_for_condition(
+        condition_predictor=_check_proxy,
+        state=proxy_state,
+        status=HTTPProxyStatus.HEALTHY,
+    )
+    first_check_time = proxy_state._last_health_check_time
+
+    # Ensure the proxy status is HEALTHY after updates.
+    assert proxy_state.status == HTTPProxyStatus.HEALTHY
+
+    # Continuously trigger update and status continue to be HEALTHY.
+    wait_for_condition(
+        condition_predictor=_check_proxy,
+        state=proxy_state,
+        status=HTTPProxyStatus.HEALTHY,
+    )
+    # Ensure the proxy status is still HEALTHY after updates.
+    assert proxy_state.status == HTTPProxyStatus.HEALTHY
+
+    # Ensure the check time have changed since the last update
+    assert first_check_time != proxy_state._last_health_check_time
+
+
+@patch("ray.serve._private.http_state.PROXY_HEALTH_CHECK_PERIOD_S", 0.1)
+def test_http_proxy_state_update_healthy_check_health_failed_once():
+    """Test calling update method on HTTPProxyState when the proxy state is HEALTHY and
+    when the check_health call failed once and succeeded for the following call.
+
+    The proxy state started with STARTING. After update is called and ready call
+    succeeded, the status will change to HEALTHY. After the next period of check_health
+    call and that check_health call failed, the status should not be set to UNHEALTHY
+    and should stay as HEALTHY. The following check_health call continue to succeed
+    and the status continue to stay as HEALTHY.
+    """
+    signal = SignalActor.remote()
+
+    @ray.remote(num_cpus=0)
+    class NewMockHTTPProxyActor:
         async def ready(self):
             return json.dumps(["mock_worker_id", "mock_log_file_path"])
 
         async def check_health(self):
-            await asyncio.sleep(100)
+            await signal.wait.remote()
 
-    proxy = MockHTTPProxyActor.options(lifetime="detached").remote()
-    state = HTTPProxyState(proxy, "alice", "mock_node_id", "mock_node_ip")
-    assert state.status == HTTPProxyStatus.STARTING
+    proxy_state = _create_http_proxy_state(proxy_actor_class=NewMockHTTPProxyActor)
+
+    # Continuously trigger update. The status should change from STARTING to HEALTHY
+    # when ready.
+    wait_for_condition(
+        condition_predictor=_check_proxy,
+        state=proxy_state,
+        status=HTTPProxyStatus.HEALTHY,
+    )
```

Comment:
Same here, if you don't expect the status to have changed from healthy during this period, I think this will only run once

Reply:
Yea, I also tried without `wait_for_condition()` none of those status will change